### ISSUE TITLE
view: Filter by alignment score (AS:i) [feature request]

### ISSUE THREAD

**Issue by sjackman** [samtools/samtools#665]
I often to filter by alignment score. It would be very helpful to me if this feature were built into `samtools view`, similar to `-q` for filtering by mapping quality.

**Comment by daviesrob** [comment]
It should be possible to add something to do this.  Or maybe we should add something to filter on the value of any tag?  A bit more complicated as it would need some sort of expression parser, but could be quite useful.

**Comment by mcshane** [comment]
@pd3 has `filter.[ch]` in bcftools that does expression parsing there. @lh3 also had `kexpr.[ch]` in https://github.com/lh3/bgt

**Comment by sjackman** [comment]
I really like the expression parser of [Miller](http://johnkerl.org/miller-releases/miller-head/doc/reference-dsl.html). It's what I'm using now in fact to filter SAM files by alignment score, after converting SAM to a format supported by Miller (e.g. TSV or JSON). It's similar to `awk`.
```sh
mlr filter '$AS >= 100'
```
http://johnkerl.org/miller-releases/miller-head/doc/reference-verbs.html#filter

**Comment by jkbonfield** [comment]
Fixed in 1.12 via PR https://github.com/samtools/samtools/pull/1346

Example syntax: `samtools view -e '[AS]>=100'`

**Comment by sjackman** [comment]
🥳 Woohoo! Thank you, James! 🏆

**Comment by cbird808** [comment]
> Fixed in 1.12 via PR #1346
> 
> Example syntax: `samtools view -e '[AS]>=100'`

This might be another issue, but is it possible to use a more complicated equation, such as 
```
samtools view -e '[AS]>=100 * ReadLength/150'
```
?

When comparing degraded DNA samples to pristine, there will be a bias for less heterozygosity in more degraded samples (insert lengths < 150 assuming pe150 sequencing) without a filter like the one I defined, which adjusts the alignment score threshold by the read length.

**Comment by jkbonfield** [comment]
Yes, you can do things like `-e '[AS]>=100*(qlen+sclen)/150'`.  You'll need to decide exactly what length you're after though.  `qlen` and `rlen` is the length of aligned query sequence and reference sequence.  `qlen` isn't exactly the length of sequence though, only those that have a placement on the genome.  You may wish to add `sclen` (number of soft-clipped bases) too, depending on what it is you're trying to work out.
